README.md
# Stimulizer

Easily work with StimulusJS `data-*` attributes in ViewComponents, Phlex templates, etc.

This gem is particularly handy in ViewComponents and Phlex when you have a 1-1 relationship between a component and a Stimulus controller, especially when using 'sidecar' controllers.

For example:

### ViewComponent
```ruby
  class Mynamespace::Deeper::ButtonsGridComponent < ApplicationComponent
  # ...
```
#### Before
```html
<div 
  data-controller="mynamespace--deeper--buttons-grid-component"
  data-mynamespace--deeper--buttons-grid-component-url-value="http://example.com"
  data-mynamespace--deeper--buttons-grid-component-color-value="#ff0000"
  data-action="click->data-mynamespace--deeper--buttons-grid-component#doSomething"
>
```

#### After
```html
<div 
  <%= stimulus(
    :controller, 
    values: {url: "http://example.com", color: "#ff0000"}, 
    action: "click->doSomething"
  )%>
>
```

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'stimulizer'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install stimulizer

## Usage
### in a classic ERB View
Include the module in your ApplicationController (or individual controllers):

```ruby
class ApplicationController < ActionController::Base
  include Stimulizer
  helper_method :stimulus
  # ...
```
NOTE that you have to add `helper_method :stimulus` to ensure the view templates can use the `stimulus` method.

### in a ViewComponent
Include this in your class:

```ruby
class MyComponent < ApplicationComponent
  include Stimulizer
  # ...
```

Render the stimulusjs data strings into your template...

```html
  <!-- use the derived controller name -->
  <%= stimulus(:controller) %>   

  <!-- specify the controller explicity, and/or add more controllers space-separated -->
  <%= stimulus(controller: "another foo")
  
  <!-- Note: 'action' is singular because it works like 'data-action' -->
  <%= stimulus(action: "click->doThing")

  <!-- but you can have multiple actions separated by spaces -->
  <%= stimulus(action: "click->doThing mouseup->otherThing")

  <!-- you can also skip the event, just like usual -->
  <%= stimulus(action: "doThing")

  <!-- 
    target for this element (if you need to also target it for other controllers, you'll have to do that manually with the old 'data-blah-target=' approach)
  -->
  <%= stimulus(target: "button")

  <!-- supports the 'classes' feature -->
  <%= stimulus(classes: {foo: "text-red-500", busy: "opacity-50 animate-spin"})
  
  <!-- supports the 'values' feature -->
  <%= stimulus(values: {url: "https://example.com"})

  <!-- supports the 'params' feature -->
  <%= stimulus(params: {foo: "bar", this_thing: "whatever"})
     
  <!-- supports the 'outlets' feature -->
  <%= stimulus(outlets: {foo: ".my-class", bar: "#widget"})     
```
... or combine them:
```
  <%= stimulus(:controller, target: "button", action: "click->doThing")
```

## Options

### `stimulize`
Use the `stimulize` method after including the module to set some config options...

#### `ignore_prefix`
Use the `ignore_prefix` option to chop off some of the namespacing if you want to shorten up those crazy-long controller filenames.
```ruby
stimulize ignore_prefix: "Components::"
```

#### `output`
Use the `output` option to change from an html-style string of data attributes (default) to a Hash of data attributes, useful in other circumstances, like tag builders and Phlex views.
```ruby
stimulize output: :hash
```

## Overriding the derived controller name
My main use case for Stimulizer is with so-called 'sidecar' javascript. That is, if I have a component in the filepath `components/foo/bar/other/button_component.rb`, I will also have a stimulus controller at `components/foo/bar/other/button_component_controller.js`. 

**In this case, Stimulizer will auto-derive the controller name properly.**

If you're not doing things this way, you may need to manually set your controller name in each component/template...

```ruby
def stimulus_controller
  "whatever--name--you-want-here"
end
```

**or...**

You can also pass the controller name directly to the `stimulus()` method. This starts to become less automagical, but still handy in reducing the repetition of those controller names.

```ruby
<div
  <%= stimulus(
    controller_name: "my--fancy--buttons", 
    action: "click->show", 
    values: {url: "example.com"}
  ) %>
>

# => <div data-controller='my--fancy--buttons' data-action='click->my--fancy--buttons#show' data-my--fancy--buttons-url-value="example.com">
```

## Using with Phlex (or if you want a hash for some other reason)
By default, Stimulizer returns a html-style string of `data-*` attributes from the `stimulus()` method. But if you're using something like Phlex templates, or if you want to use this with a tag-building method, you might want the original Hash instead. You have two options:

### 1. Class-level configuration
```ruby
module Views
  class ApplicationView < Phlex::View
    include Stimulizer
    stimulize output: :hash
```

### 2. Specifically ask for a hash with `stimulus_hash` instead
```ruby
stimulus_hash(:controller, action: "click->doThis")
```

## Phlex: Destructuring for the win!
Thanks to [@joeldrapper](https://github.com/joeldrapper) for suggesting this and for inspiring the idea of this gem. You can use destructuring to great effect in Phlex views:

```ruby
div(**stimulus(:controller, action: "click->show"), class: "text-red-500") do
  # ...
```
*Note that this example is assuming you've turned on `:hash` mode as above. You could also use `**stimulus_hash()` in the same way.*

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/stimulizer. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/stimulizer/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Stimulizer project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/stimulizer/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2022-11-03

- Initial release

## [0.2.1] - 2022-11-29

- Adds support for `outlets` in Stimulus 3.2.0


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "stimulizer"

require "minitest/autorun"


test/stimulizer_test.rb
# frozen_string_literal: true

require "test_helper"

class StimulizerTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::Stimulizer::VERSION
  end

  def test_it_does_something_useful
    assert false
  end
end


lib/stimulizer.rb
# frozen_string_literal: true

require "stimulizer"

require "active_support/concern"
require "dry-configurable"
require "lucky_case"

require_relative "stimulizer/version"
module Stimulizer
  extend Dry::Configurable
  # extend ActiveSupport::Concern

  class Error < StandardError; end

  def self.included(klass)
    klass.extend(ClassMethods)
  end

  module ClassMethods
    @@stimulizer_opts = {output: :html_string}

    def _stimulizer_opts
      @@stimulizer_opts
    end

    def stimulize(hash)
      @@stimulizer_opts.merge!(hash.deep_symbolize_keys)
    end
  end

  def stimulus_controller
    build_stimulus_controller
  end

  def stimulus_hash(*args)
    if args.size == 1 && !args.first.is_a?(Hash)
      build_stimulus_hash(args.first)
    elsif args.size == 1 && args.first.is_a?(Hash)
      build_stimulus_hash(**args.first)
    elsif args.size == 2 && args[1..].all? { |a| a.is_a?(Hash) }
      build_stimulus_hash(args[0], **args[1..].first)
    else
      raise ArgumentError, "wrong number of arguments (given #{args.size}, expected 1..2)"
    end
  end

  def stimulus(*args)
    output = stimulus_hash(*args)
    return output if _stimulizer_opts[:output] == :hash

    output.to_a
      .map { |k, v| %(#{k.to_s.strip}="#{v.to_s.strip}") }
      .join(" ")
      .squish
      .strip
  end

  private

  def _stimulizer_opts
    self.class._stimulizer_opts
  end

  # EXAMPLE: My::Fancy::FooThing => "my--fancy--foo-thing"
  def build_stimulus_controller
    klass = self.class.name
    klass = klass.gsub(_stimulizer_opts[:ignore_prefix], "") if _stimulizer_opts[:ignore_prefix].present?
    klass.underscore.dasherize.gsub("/", "--")
  end

  # USAGE:
  #   <%= stimulus(:controller)   <-- USES THE DEFAULT CONTROLLER
  #   <%= stimulus(controller: "my-fancy-controller")
  #   <%= stimulus(action: "click->doThing")
  #   <%= stimulus(action: "click->doThing mouseup->otherThing")
  #   <%= stimulus(action: "doThing") <-- default event for element
  #   <%= stimulus(target: "button")
  #   <%= stimulus(classes: {foo: "text-red-500", busy: "opacity-50 animate-spin"})
  #   <%= stimulus(values: {url: "https://example.com"})
  #   <%= stimulus(params: {foo: "bar", this_thing: "whatever"})
  #
  # ... or combine them:
  #
  #   <%= stimulus(:controller, target: "button", action: "click->doThing")
  def build_stimulus_hash(default_controller = false, controller_name: nil, controller: nil, target: nil, action: nil, params: nil, values: nil, classes: nil, outlets: nil)
    raise ArgumentError(":controller_name specified, but blank") if controller_name&.blank?
    raise ArgumentError(":controller specified, but blank") if controller&.blank?
    raise ArgumentError(":target specified, but blank") if target&.blank?
    raise ArgumentError(":action specified, but blank") if action&.blank?
    raise ArgumentError(":params specified, but blank") if params&.blank?
    raise ArgumentError(":values specified, but blank") if values&.blank?
    raise ArgumentError(":classes specified, but blank") if classes&.blank?
    raise ArgumentError(":outlets specified, but blank") if outlets&.blank?

    {}.tap do |hash|
      hash[:"data-controller"] = ""

      local_controller_name = stimulus_controller

      if default_controller.to_s.downcase.to_sym == :controller
        hash[:"data-controller"] += " #{stimulus_controller}"
      elsif controller_name.present?
        hash[:"data-controller"] += " #{controller_name}"
        local_controller_name = controller_name
      end

      hash[:"data-controller"] += " #{controller}" if controller

      hash.delete(:"data-controller") if hash[:"data-controller"].blank?

      if action
        actions = action.squish.split(" ").map do |str|
          next if str.blank?

          arr = str.split("->")
          event, function = arr.last.nil? ? [nil, arr.first] : ["#{arr.first}->", arr.last]

          if function.include?("#")
            "#{event}#{function}"
          else
            "#{event}#{local_controller_name}##{function}"
          end
        end.compact

        hash[:"data-action"] = actions.join(" ")
      end

      params&.each do |key, value|
        hash[:"data-#{local_controller_name}-#{LuckyCase.dash_case(key.to_s)}-param"] = value
      end

      values&.each do |key, value|
        hash[:"data-#{local_controller_name}-#{LuckyCase.dash_case(key.to_s)}-value"] = value
      end

      classes&.each do |key, value|
        hash[:"data-#{local_controller_name}-#{LuckyCase.dash_case(key.to_s)}-class"] = value
      end

      outlets&.each do |key, value|
        hash[:"data-#{local_controller_name}-#{LuckyCase.dash_case(key.to_s)}-outlet"] = value
      end

      hash[:"data-#{local_controller_name}-target"] = target if target
    end
  end
end
